test/test_virtual_device.py

```diff
@@ -0,0 +1,84 @@
+import os
+import sys
+
+import unittest
+import numpy as np
+
+import torch
+import torch_xla.core.xla_model as xm
+import torch_xla.experimental.xla_sharding as xs
+from torch_xla.experimental.xla_sharded_tensor import XLAShardedTensor
+from torch_xla.experimental.pjrt import using_pjrt
+
+
+@unittest.skipIf(not using_pjrt() or xm.get_xla_supported_devices("GPU"),
+                 f"Requires PJRT_DEVICE set to `TPU` or `CPU`.")
+class XlaVirtualDeviceTest(unittest.TestCase):
+
+  n_devices = 0
+  device_ids = None
+
+  @classmethod
+  def setUpClass(cls):
+    os.environ["XLA_USE_SPMD"] = "0"
+    cls.n_devices = len(xm.get_xla_supported_devices())
+    os.environ["XLA_USE_SPMD"] = "1"
+    cls.device_ids = np.array(range(cls.n_devices))
+
+  def _get_mesh(self, mesh_shape, device_ids=None):
+    if device_ids is None:
+      device_ids = self.device_ids
+    assert len(device_ids) == self.n_devices
+    return xs.Mesh(device_ids, mesh_shape)
+
+  def test_mark_sharding(self):
+    partition_spec = (0, 1)
+    xt1 = torch.tensor([[1, 2, 3, 4, 5, 6, 7, 8]],
+                       dtype=torch.float,
+                       device=xm.xla_device())
+    xst1 = xs.mark_sharding(xt1, self._get_mesh((1, self.n_devices)),
+                            partition_spec)
+    self.assertTrue(isinstance(xst1, XLAShardedTensor))
+
+  def test_ops_on_host(self):
+    t1 = torch.tensor([[1, 2, 3, 4, 5, 6, 7, 8]],
+                      dtype=torch.float,
+                      device=xm.xla_device())
+    t2 = torch.tensor([[8, 7, 6, 5, 4, 3, 2, 1]],
+                      dtype=torch.float,
+                      device=xm.xla_device())
+    t3 = t1 + t2
+    self.assertTrue(isinstance(t3, torch.Tensor))
+
+  def test_ops_on_devices_with_implicit_sharding(self):
+    partition_spec = (0, 1)
+    t1 = torch.tensor([[1, 2, 3, 4, 5, 6, 7, 8]],
+                      dtype=torch.float,
+                      device=xm.xla_device())
+    t2 = torch.tensor([[8, 7, 6, 5, 4, 3, 2, 1]],
+                      dtype=torch.float,
+                      device=xm.xla_device())
+    xst1 = xs.mark_sharding(t1, self._get_mesh((1, self.n_devices)),
+                            partition_spec)
+    t3 = xst1 + t2
```

Comment:
Can we do `t3=t1+t2` instead? We may make the `mark_sharding` API in-place in the future instead of returning `xst1`. Also, we should compare the result for correctness in value -- e.g., comparing with `[9, 9, 9, 9, 9,9 ,9 9]` and with or without the virtual device optimzation in this case. 

This applies to the other test cases here.  As a side note, we could place the entire test class in test_xla_sharding.py and reuse some of the test constructs, so we can group all the spmd related unit tests in one file?

Reply:
Done